sdk/tests/core/test_az_mqtt5_connection.c

```diff
@@ -434,28 +668,917 @@ static void test_az_mqtt5_connection_open_idle_success(void** state)
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_connected(&mock_mqtt);
+
   assert_int_equal(
       az_mqtt5_inbound_connack(
           &mock_mqtt,
-          &(az_mqtt5_connack_data){ .connack_reason = CONNACK_SUCCESS,
+          &(az_mqtt5_connack_data){ .connack_reason = AZ_MQTT5_CONNACK_SUCCESS,
                                     .tls_authentication_error = false }),
       AZ_OK);
-  assert_int_equal(ref_conn_req, 1);
+
   assert_int_equal(ref_conn_rsp_ok, 2);
+  assert_int_equal(ref_conn_error, 1);
+
+  _az_mqtt5_connection_faulted_check(&test_connection);
+}
+
+// Testing that when connected, a connect response error from the broker leads to a faulted state.
+static void test_az_mqtt5_connection_connected_connect_rsp_err_failure(void** state)
+{
+  (void)state;
+  _reset_refs();
+  az_mqtt5_connection test_connection;
+  az_mqtt5 mock_mqtt;
+  az_mqtt5_options mock_mqtt_options = { 0 };
+  _az_event_client mock_client_1;
+  _az_event_client mock_client_2;
+  _az_hfsm mock_client_hfsm_1;
+  _az_hfsm mock_client_hfsm_2;
+
+  _az_mqtt5_connection_setup(
+      &test_connection,
+      &test_connection_options,
+      &mock_mqtt,
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_connected(&mock_mqtt);
+
+  assert_int_equal(
+      az_mqtt5_inbound_connack(
+          &mock_mqtt,
+          &(az_mqtt5_connack_data){ .connack_reason = AZ_MQTT5_CONNACK_UNSPECIFIED_ERROR,
+                                    .tls_authentication_error = false }),
+      AZ_OK);
+
+  assert_int_equal(ref_conn_rsp_ok, 2);
+  assert_int_equal(ref_conn_rsp_err, 0);
+  assert_int_equal(ref_conn_error, 1);
+
+  _az_mqtt5_connection_faulted_check(&test_connection);
+}
+
+// Testing that when connected, a disconnect request leads to a disconnecting state.
+static void test_az_mqtt5_connection_connected_disconnect_rsp_success(void** state)
+{
+  (void)state;
+  _reset_refs();
+  az_mqtt5_connection test_connection;
+  az_mqtt5 mock_mqtt;
+  az_mqtt5_options mock_mqtt_options = { 0 };
+  _az_event_client mock_client_1;
+  _az_event_client mock_client_2;
+  _az_hfsm mock_client_hfsm_1;
+  _az_hfsm mock_client_hfsm_2;
+
+  _az_mqtt5_connection_setup(
+      &test_connection,
+      &test_connection_options,
+      &mock_mqtt,
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_connected(&mock_mqtt);
+
+  will_return(__wrap_az_platform_clock_msec, 0);
+  will_return(__wrap_az_platform_clock_msec, 0);
+  will_return(__wrap_az_mqtt5_outbound_connect, AZ_OK);
```

Comment:
Why does the disconnect trigger an outbound_connect?

Reply:
When in a connected state, if we receive a disconnect from the broker we will attempt to reconnect. Good catch though, the comment was not indicative of that